tls/understand-tls/understand-tls.tex

```diff
@@ -0,0 +1,275 @@
+in a very similar way  to that described above. The primary difference is that
+the client will also have to create a private/public key pair and obtain a
+digital certificate from a CA that the server trusts. 
+
+\section{Ciphersuites}
+
+SSL/TLS itself does not mandate the use of any particular cryptographic 
+algorithms. Instead it provides a framework for combining different algorithms 
+together and enabling the client and server to negotiate between them which 
+combination of algorithms will be used to protect messages that are exchanged. 
+A group of algorithms combined in this way is known as a \emph{ciphersuite}. 
+There are many different ciphersuites that are available\footnote{At the time 
+of writing OpenSSL 1.1.0 supports 168 different ciphersuites}. Each
+ciphersuite identifies a set of algorithms that it will use to satsify the 
+following cryptographic primitives:
+\begin{itemize}
+\item Authentication. What algorithm will be used to digitally sign various 
+aspects of the communication in order to establish and verify the identify of
+the parties (either just the server, or both the server and the client). The 
+algorithm used here will be the same one used to generate the public/private 
+key pair associated with the digital certificate. Examples of common algorithms 
+include RSA, DSA and ECDSA.
+\item Key Exchange. Typically a new encryption key will be generated for each
+connection. Do not confuse this encryption key with the public/private key pair
+associated with the digital certificate. The encryption key must be shared
+between both ends of the communication and must be \emph{private} to prevent an
+eavesdropper from being able to decrypt messages. Key exchange algorithms solve 
+the problem of how two parties will agree on a key without enabling an 
+eavedropper to work out what it is. Examples of algorithms that are used for 
+this purpose include RSA, DH and ECDH.\footnote{You will also very
+frequently come across the so-called ``ephemeral'' variants of DH and ECDH
+known as DHE and ECHDE respectively.}
+\item Encryption. Having established a shared encryption key, the two 
+communicating parties can start to protect their communications from 
+eavesdroppers by encrypting the data that they exchange. Examples of common 
+encryption algorithms include AES, CAMELLIA and ChaCha20.
+\item Integrity. The algorithm used to protect communications from being 
+tampered with by an attacker. Often the algorithm used here will be one known 
+as HMAC combined with a message digest algorithm such as SHA256 or SHA512. A
+message digest is simply a secure ``hash'' function. They take arbitrary 
+length input data and output a fixed length ``hash'' value that exhibits 
+certain security properties (including that it is not possible to derive the 
+input data from the hash output). Alternatively, integrity could be provided by
+a \emph{mode} of the underlying  encryption cipher. Modes are a complex topic,
```

Comment:
Should *mode* be notified in the 'Encryption' section above? 'Modes' sounds more like to use defined due to symmetric ciphers....

Reply:
Right - but in the context of SSL/TLS, we're only really interested in whether the mode provides integrity or not. So in my mind it seems to sit better here.

Reply:
Okay, that's also rational.